File: app/src/main/java/com/example/cameraaccess/service/UploadStateBus.kt
package com.example.cameraaccess.service

import com.example.cameraaccess.monitoring.AppHealthMonitor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

data class UploadState(
    val currentRecordingId: Long? = null,
    val progress: Float = 0f,
    val stage: String = "",
    val error: String? = null,
    val isUploading: Boolean = false,
    val finishedRecordingId: Long? = null
)

object UploadStateBus {

    private val _state = MutableStateFlow(UploadState())
    val state = _state.asStateFlow()

    fun startUpload(id: Long) {
        AppHealthMonitor.recordMetric(
            "upload_state_start",
            mapOf("recording_id" to id.toString())
        )
        _state.update {
            it.copy(
                currentRecordingId = id,
                progress = 0f,
                stage = "Preparing...",
                error = null,
                isUploading = true,
                finishedRecordingId = null
            )
        }
    }

    fun setStage(stage: String) {
        AppHealthMonitor.recordMetric(
            "upload_stage_changed",
            mapOf("stage" to stage.take(50))
        )
        _state.update {
            it.copy(
                stage = stage,
                isUploading = true
            )
        }
    }

    fun updateProgress(progress: Float) {
        _state.update {
            it.copy(
                progress = progress,
                isUploading = true
            )
        }
    }

    fun finishSuccess(id: Long) {
        AppHealthMonitor.recordMetric(
            "upload_state_success",
            mapOf("recording_id" to id.toString())
        )
        _state.update {
            it.copy(
                currentRecordingId = null,
                progress = 1f,
                error = null,
                isUploading = false,
                finishedRecordingId = id
            )
        }
    }

    fun error(message: String) {
        AppHealthMonitor.recordMetric(
            "upload_state_error",
            mapOf("message" to message.take(80))
        )
        _state.update {
            it.copy(
                error = message,
                isUploading = false
            )
        }
    }

    fun reset() {
        _state.value = UploadState()
    }
}

File: TauResearch/Services/UploadStateBus.swift
// UploadStateBus.swift
// Mirrors service/UploadStateBus.kt — global @Observable singleton driving the dashboard's
// upload progress UI.

import Foundation
import Observation

struct UploadServiceState: Equatable {
    var currentRecordingId: UUID?
    var progress: Float = 0
    var stage: String = ""
    var error: String?
    var isUploading: Bool = false
    var finishedRecordingId: UUID?
}

@Observable
final class UploadStateBus {

    static let shared = UploadStateBus()

    private(set) var state = UploadServiceState()

    private init() {}

    func startUpload(id: UUID) {
        AppHealthMonitor.shared.recordMetric(name: "upload_state_start",
                                             attributes: ["recording_id": id.uuidString])
        state = UploadServiceState(
            currentRecordingId: id,
            progress: 0,
            stage: "Preparing…",
            error: nil,
            isUploading: true,
            finishedRecordingId: nil
        )
    }

    func setStage(_ stage: String) {
        AppHealthMonitor.shared.recordMetric(name: "upload_stage_changed",
                                             attributes: ["stage": String(stage.prefix(50))])
        state.stage = stage
        state.isUploading = true
    }

    func updateProgress(_ progress: Float) {
        state.progress = progress
        state.isUploading = true
    }

    func finishSuccess(id: UUID) {
        AppHealthMonitor.shared.recordMetric(name: "upload_state_success",
                                             attributes: ["recording_id": id.uuidString])
        state = UploadServiceState(
            currentRecordingId: nil,
            progress: 1,
            stage: "Done",
            error: nil,
            isUploading: false,
            finishedRecordingId: id
        )
    }

    func error(_ message: String) {
        AppHealthMonitor.shared.recordMetric(name: "upload_state_error",
                                             attributes: ["message": String(message.prefix(80))])
        state.error = message
        state.isUploading = false
    }

    func reset() {
        state = UploadServiceState()
    }
}
